windows/lcow_runtime.go

```diff
@@ -0,0 +1,426 @@
+	/////////	return nil, errors.Wrapf(err, "failed to create sandbox layer for %s: layers: %#v, driverInfo: %#v",
+	/////////		id, layerFolders, di)
+	/////////}
+	/////////defer func() {
+	/////////	if err != nil {
+	/////////		removeLayer(ctx, conf.LayerFolderPath)
+	/////////	}
+	/////////}()
+	/////////
+	/////////if err = hcsshim.ActivateLayer(di, id); err != nil {
+	/////////	return nil, errors.Wrapf(err, "failed to activate layer %s", conf.LayerFolderPath)
+	/////////}
+	/////////
+	/////////if err = hcsshim.PrepareLayer(di, id, layerFolders); err != nil {
+	/////////	return nil, errors.Wrapf(err, "failed to prepare layer %s", conf.LayerFolderPath)
+	/////////}
+	/////////
+	/////////conf.VolumePath, err = hcsshim.GetLayerMountPath(di, id)
+	/////////if err != nil {
+	/////////	return nil, errors.Wrapf(err, "failed to getmount path for layer %s: driverInfo: %#v", id, di)
+	/////////}
+
+	var (
+		conf *hcsshim.ContainerConfig
+		nsid = namespace + "-" + id
+	)
+	if conf, err = newLinuxConfig(ctx, hcsshimOwner, nsid, spec); err != nil {
+		return nil, err
+	}
+
+	if err := r.CreateReadWrite(id, spec.Windows.LayerFolders[0]); err != nil {
+		return nil, err
+	}
+	conf.LayerFolderPath = r.dir(id)
+	defer func() {
+		if err != nil {
+			removeLinuxLayer(ctx, conf.LayerFolderPath)
+		}
+	}()
+
+	// TODO: remove this once we have an LCOW snapshotter
+	// Store the LayerFolder in the db so we can clean it if we die
+	if err = r.db.Update(func(tx *bolt.Tx) error {
+		s := newLayerFolderStore(tx)
+		return s.Create(nsid, conf.LayerFolderPath)
+	}); err != nil {
+		return nil, err
+	}
+	defer func() {
+		if err != nil {
+			if dbErr := r.db.Update(func(tx *bolt.Tx) error {
+				s := newLayerFolderStore(tx)
+				return s.Delete(nsid)
+			}); dbErr != nil {
+				log.G(ctx).WithField("id", id).
+					Error("failed to remove key from metadata")
+			}
+		}
+	}()
+
+	ctr, err := hcsshim.CreateContainer(nsid, conf)
+	if err != nil {
+		time.Sleep(15 * time.Second)
+		return nil, errors.Wrapf(err, "hcsshim failed to create task")
+	}
+	defer func() {
+		if err != nil {
+			ctr.Terminate()
+			ctr.Wait()
+			ctr.Close()
+		}
+	}()
+
+	if err = ctr.Start(); err != nil {
+		return nil, errors.Wrap(err, "hcsshim failed to spawn task")
+	}
+
+	t := &task{
+		id:                id,
+		namespace:         namespace,
+		pid:               pid,
+		io:                pset,
+		status:            runtime.CreatedStatus,
+		spec:              spec,
+		processes:         make(map[string]*process),
+		publisher:         r.publisher,
+		rwLayer:           conf.LayerFolderPath,
+		pidPool:           r.pidPool,
+		isWindows:         false,
+		hyperV:            true,
+		hcsContainer:      ctr,
+		terminateDuration: createOpts.TerminateDuration,
+	}
+	r.tasks.Add(ctx, t)
+
+	var rootfs []*containerdtypes.Mount
+	for _, l := range append([]string{t.rwLayer}, spec.Windows.LayerFolders...) {
```

Comment:
Should this not happen client side?

Reply:
I guess we can do this once we have an LCOW snapshotter.

Reply:
Yeah, this is just constructing something for the metadata due to missing the snapshotter. This will be done in the snapshotter eventually.